### PR TITLE
<SortableList/> - disable drag for specific list items

### PR DESCRIPTION
**Pull request by Wo1v3r**
### What changed
Added disabled prop to draggable item
...

### Why it changed
Needed in order to have static elements in a DnD list for Price Quotes (Table header) or Trash like components
...

---

- [ ] Change is tested
- [ ] Change is documented
- [ ] Build is green
- [ ] Change of UX is approved by [wuwa](https://github.com/wuwa) or [milkyfruit](https://github.com/milkyfruit)

### CODE REVIEW

**Comment by Wo1v3r** [comment]
@Faradey27  Andrii pls have a look, maybe you want to move it to another file e.g `DraggableDisabled` seperately ?

File: src/DragAndDrop/Draggable/Draggable.js [hunk 1]
```diff
@@ -7,9 +7,16 @@ import DraggableTarget from './components/DraggableTarget';
 
 export class Draggable extends WixComponent {
   render() {
-    return (
+    return this.props.disabled ? (
+      this.props.renderItem({
+        isPlaceHolder: false,
+        isPreview: false,
+        connectHandle: _ => _,
```
Comment (shlomitc) [inline src/DragAndDrop/Draggable/Draggable.js]: use `lodash.noop` here

File: stories/DragAndDrop/SortableList/SingleAreaList.js [hunk 2]
```diff
@@ -26,6 +26,11 @@ export default class SingleAreaList extends React.Component {
       {
         id: 'd',
         text: 'Item 4'
+      },
+      {
+        id: 'e',
+        text: 'Disabled Item 5',
+        disabled: true
```
Comment (shlomitc) [inline stories/DragAndDrop/SortableList/SingleAreaList.js]: how will this affect the CSS of the non-dragged component?
for example - if there is a pointer cursor before drag - it should now not be set to the initial one.
I don't see this is propagated to the renderItem function, so I wonder how it affects it.

Comment (Wo1v3r) [reply stories/DragAndDrop/SortableList/SingleAreaList.js]: was thinking of adding another `isDisabled` but it is redundant as all of the above are passed via props therefore `disabled` would already be passed to the `renderItem` function, I can add relevant style change in the example. not sure how that will be clear in the API though (that all props pass through)

Comment (shlomitc) [reply stories/DragAndDrop/SortableList/SingleAreaList.js]: This looks to me like one of those values we should pass in a straightforward way like we pass the `id` - will also be easier to document it and understand the flow.

yes, style changes are needed. I think the current common styles lack some edge case, this one for example.

File: src/DragAndDrop/Draggable/Draggable.js (hunk 1 above)
Comment (shlomitc) [reply src/DragAndDrop/Draggable/Draggable.js]: We use lodash.noop anyway in many other places so it’s not a new dependency. It also makes sure we do not create new functions accidentally and easier to explain required changes in PRs. So lodash.noop is exactly like the functions you wrote, I see no reason why not to use it

Comment (argshook) [reply src/DragAndDrop/Draggable/Draggable.js]: it is used here in Draggable and also in Tooltip, it's just two, not many :)

this doesn't matter much but i wouldn't use 3rd party helpers for things as trivial as this

Comment (Wo1v3r) [reply src/DragAndDrop/Draggable/Draggable.js]: Um ok originally I used identity there in connectHandle to return the handle unconnected, but I see it's noop now (e.g `() => undefined`).

I'll update in the PR with lodash.noop as it's already being used and will be easier to find and replace if @argshook  wishes (I agree it better be under utils than 3rd party [for one liners])

File: stories/DragAndDrop/SortableList/SingleAreaList.js (hunk 2 above)
Comment (Wo1v3r) [reply stories/DragAndDrop/SortableList/SingleAreaList.js]: Cool np will update with example and with something similar to id;  <{...restProps} id={id} disabled={disabled} />

**Comment by Wo1v3r** [comment]
@shlomitc  @Faradey27  @argshook  I updated the PR;
- used `lodash/identity` to also keep the handle in place etc, if user of Draggable will want otherwise, he can avoid rendering using the disabled prop, IMO the default should be to keep the component intact mostly
- passing `item` and `id` explicitly to `renderItem` to be more clear
- added styles to the example (colors are WSR but not sure they match styleguide)

File: src/DragAndDrop/Draggable/Draggable.js [hunk 3]
```diff
@@ -1,15 +1,26 @@
...
 
 import WixComponent from '../../BaseComponents/WixComponent';
 import DraggableSource from './components/DraggableSource';
 import DraggableTarget from './components/DraggableTarget';
 
 export class Draggable extends WixComponent {
   render() {
-    return (
+    const { item, id } = this.props;
+    return this.props.disabled ? (
```
Comment (Faradey27) [inline src/DragAndDrop/Draggable/Draggable.js]: Draggable should not know internal api of DragSource, better to keep such logic inside of a DragSource

File: src/DragAndDrop/Draggable/Draggable.js [hunk 4]
```diff
@@ -39,7 +50,9 @@ Draggable.propTypes = {
   /** callback for drag start */
   onDragStart: PropTypes.func,
   /** callback for drag end */
-  onDragEnd: PropTypes.func
+  onDragEnd: PropTypes.func,
+  /** Disabled Drag and Drop Item */
+  disabled: PropTypes.bool
```
Comment (Faradey27) [inline src/DragAndDrop/Draggable/Draggable.js]: maybe draggable? or fixed?

File: stories/DragAndDrop/SortableList/SingleAreaList.js [hunk 5]
```diff
@@ -10,48 +10,61 @@ import styles from './SingleAreaList.scss';
...
+          id: 'e',
+          text: 'Disabled Item 5',
+          disabled: true
+        }
+      ]
+    };
   }
 
-  handleDrop = ({removedIndex, addedIndex}) => {
+  handleDrop = ({ removedIndex, addedIndex }) => {
```
Comment (Faradey27) [inline stories/DragAndDrop/SortableList/SingleAreaList.js]: wsr eslint will ask to remove spaces

**Comment by Faradey27** [comment]
@Wo1v3r ping, let finish PR) the feature may be useful for a lot of people

**Comment by Wo1v3r** [comment]
@Faradey27 ok pushing in a moment

**Comment by shlomitc** [comment]
1. @Faradey27 - are you satisfied by this solution?
2. @Wo1v3r build is failing for lint errors and it is conflicting now with master, can you please fix it?

**Comment by Wo1v3r** [comment]
@shlomitc  hopefully will get to it this thursday

**Comment by Faradey27** [comment]
@shlomitc I think it enough to merge, but still I think <Draggable/> should not know what is renderItem

**Comment by shlomitc** [comment]
I updated the branch (pushed-force), so notice that.
I also moved the example to the SortableList section.

Look however that the solution is a bit problematic.

for example, try to drag item 2 below item 3.
This will not trigger the drop area below it until it goes over item 4.

I think this is not the best user experience. @Faradey27 any other suggestions?

**Comment by Faradey27** [comment]
> I updated the branch (pushed-force), so notice that.
> I also moved the example to the SortableList section.
> 
> Look however that the solution is a bit problematic.
> 
> for example, try to drag item 2 below item 3.
> This will not trigger the drop area below it until it goes over item 4.
> 
> I think this is not the best user experience. @Faradey27 any other suggestions?

sounds bad, so, I think when we say "disabled" we mean that we still want to drop on the item, but do not want to allow to drag it.

So, maybe let just prevent DragSource wrap? like inside of DragSource we will just call and return renderItem without using d&d HOCs(currenty in PR we are doing this inside of the Draggable). Also, maybe then let rename prop `disabled` to `draggable` or smth like this

**Comment by Wo1v3r** [comment]
@Faradey27  I'm thinking of retracting this PR for now. the need came from a table use case, but with all the behaviour divs attached along the way, it makes it a tad problematic to use in a table.

I'm trying to implement on my own so we can discuss it (regarding the actual use case)

--

In general, I agree that it should be split into two different components (DragSource, DropTarget) and export as such (e.g no "Draggable" only)
then each of the HOCs will have its' own `disabled` prop with matching behaviour.
WDYT?

**Pull request closed without merging**